requests-for-comments/2023-05-30-clickhouse-as-only-store.md

```diff
@@ -0,0 +1,35 @@
+# Request for comments: Clickhouse as our only data store
+
+## Problem statement
+
+- Having two separate places where we store essentially the same data (person/person_distinct_id/groups) means we often have inconsistencies
+- Our plans with the data warehouse are to allow customers to model all of their entities (persons/groups) in SQL in Clickhouse. For this to work we need to be able to read everything everywhere from clickhouse
+- When postgres goes down due to load from this data, it takes down the entire app
+
+## Success criteria
+*How do we know if this is successful (i.e. metrics, customer feedback), what's out of scope, whats makes this ambitious?*
+We no longer store any person/groups data on postgres
+
+## Context
+I think we could revisit some of our assumptions on what our system needs to be able to do this. For example, I think an hour or days delay for analytics, as long as we still have a real time view of events coming in, even if they aren't enriched with person data.
+
+## Design 
+
+There are a couple of places we use postgres
+
+**Ingestion**
+Options
+- Read person properties/groups/distinct ids from Clickhouse, with some kind of cache to avoid overloading
+- Materialized view that joins the person properties onto the table, perhaps at some interval like an hour or day.
+
+**Decide/feature flags**
```

Comment:
flyby:

it feels like it's ok to run postgres as a service for just feature flags and to write/read there for pure speed reasons. in the same way we may need to create (some specialist service) for (some specialist product) in future.  it just happens to be its own DB.

 it'd mean some minor inconsistencies between persons for feature flags versus everything else, but that's worth the tradeoff for performance here. it feels worse to have inconsistencies in analytics data / business reporting data, hence the other ideas here i think make more sense